### DISCUSSION TITLE
v3.0.0

### DISCUSSION THREAD

**Discussion by rawilk** [rawilk/laravel-printing#29] [Releases]
### Added
- Add driver method for retrieving print jobs (**Breaking Change** to driver contract)
- Add driver method for retrieving a specific print job (**Breaking Change** to driver contract)
- Add driver method for retrieving a specific printer's print jobs (**Breaking Change** to driver contract)
- Add driver method for retrieving a specific print job on a specific printer (**Breaking Change** to driver contract)
- Add `printer()` method on PrintNode driver printer to access underlying PrintNode printer instance
- Add `job()` method on PrintNode driver print job to access underlying PrintNode print job instance
- Add a `printer` property on the PrintNode driver PrintJob class to access the printer instance

### Changed
- **Breaking Change:** Rename driver method `find()` to `printer()` for finding a specific printer
- **Breaking Change:** Add required `$limit`, `$offset`, and `$dir` pagination params to driver `printers()` method
- **Breaking Change:** Add `null|Carbon` return type to `PrintJob` contract `date()` method signature
- Write our own internal api wrapper for PrintNode driver instead of relying on package `printnode/printnode-php` (available via `app(\Rawilk\Printing\Api\PrintNode\PrintNode::class)`)
- Make `\Rawilk\Printing\Printing` macroable
- Make `Rawilk\Printing\PrintTask` macroable
- Make `Rawilk\Printing\Drivers\PrintNode\PrintNode` macroable
- Make `Rawilk\Printing\Drivers\Cups\Cups` macroable
- Make each concrete instance of `\Rawilk\Printing\Contracts\Printer` and `\Rawilk\Printing\Contracts\PrintJob` macroable
- Make `\Rawilk\Printing\Receipts\ReceiptPrinter` macroable

### Fixed
- Make `\Rawilk\Printing\Drivers\PrintNode\Entity\Printer` compatible with implemented `JsonSerializable` interface
- Return a given PrintNode driver printer instance's jobs via the `jobs()` method

## New Contributors
* @LooxisDev made their first contribution in https://github.com/rawilk/laravel-printing/pull/23

**Full Changelog**: https://github.com/rawilk/laravel-printing/compare/2.0.0...v3.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/rawilk/laravel-printing/releases/tag/v3.0.0'>v3.0.0</a>.</em>

**Comment by Batoua** [comment]
Hos to solve Cannot instantiate abstract class Rawilk\Printing\PrintTask?
I want some help

This is my code

use \Rawilk\Printing\PrintTask;
use \Rawilk\Printing\Receipts\ReceiptPrinter;
public function xprinter(Request $request)
    {
        try {
            $text = (string) (new ReceiptPrinter)
            ->centerAlign()
            ->text('My heading')
            ->leftAlign()
            ->line()
            ->twoColumnText('Item 1', '2.00')
            ->twoColumnText('Item 2', '4.00')
            ->feed(2)
            ->centerAlign()
            ->barcode('1234')
            ->cut();

            $printJob= new PrintTask;
            $printJob->printer("71138085");
            $printJob->content($text);
            $printJob->send();
        } catch(Exception $e) {
            echo "Couldn't print to this printer: " . $e -> getMessage() . "\n";
        }
    }

**Comment by rawilk** [comment]
You're trying to create a new instance of an abstract class, which you can't do in PHP. You should be using the facades instead for creating a new print job.

Please refer to the [docs](https://randallwilk.dev/docs/laravel-printing/v3/basic-usage/basic-usage) for more information.

**Reply by Batoua** [reply]
Thank, this code past on my local server but on my online webserver, i have this error message
Target [Rawilk\Printing\Contracts\Driver] is not instantiable while building [Rawilk\Printing\Printing].

use \Rawilk\Printing\Facades\Printing;
use \Rawilk\Printing\Receipts\ReceiptPrinter;
public function xprinter(Request $request, Printing $printing)
    {
        try {
            $receipt = (string) (new ReceiptPrinter)
            ->centerAlign()
            ->fontSize(3, 5)
            ->text('My heading')
            ->fontSize(1, 1)
            ->leftAlign()
            ->line()
            ->twoColumnText('Item 1', '2.00')
            ->twoColumnText('Item 2', '4.00')
            ->feed(2)
            ->centerAlign()
            ->barcode('1234')
            ->cut();

           $printing::newPrintTask()
            ->printer("71138085")
            ->content($receipt)
            ->send();
        } catch(Exception $e) {
            echo "Couldn't print to this printer: " . $e -> getMessage() . "\n";
        }
    }
